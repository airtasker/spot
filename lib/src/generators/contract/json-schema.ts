import assertNever from "assert-never";
import * as YAML from "js-yaml";
import { Api, normalizedObjectType, Type, Types } from "../../models";
import compact = require("lodash/compact");

export function generateJsonSchema(api: Api, format: "json" | "yaml") {
  const contract = jsonSchema(api);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract);
    default:
      throw assertNever(format);
  }
}

export function jsonSchema(api: Api): JsonSchema {
  return {
    $schema: "http://json-schema.org/draft-07/schema#",
    definitions: Object.entries(api.types).reduce(
      (acc, [typeName, type]) => {
        acc[typeName] = jsonTypeSchema(api.types, type);
        return acc;
      },
      {} as { [typeName: string]: JsonSchemaType }
    )
  };
}

export function jsonTypeSchema(types: Types, type: Type): JsonSchemaType {
  switch (type.kind) {
    case "void":
    case "null":
      return {
        type: "null"
      };
    case "boolean":
      return {
        type: "boolean"
      };
    case "boolean-constant":
      return {
        type: "boolean",
        const: type.value
      };
    case "date":
    case "date-time":
    case "string":
      return {
        type: "string"
      };
    case "string-constant":
      return {
        type: "string",
        const: type.value
      };
    case "number":
    case "int32":
    case "int64":
    case "float":
    case "double":
      return {
        type: "number"
      };
    case "integer-constant":
      return {
        type: "integer",
        const: type.value
      };
    case "object":
      return Object.entries(normalizedObjectType(types, type)).reduce(
        (acc, [name, type]) => {
          if (type.kind === "optional") {
            type = type.optional;
          } else {
            acc.required.push(name);
          }
          const schemaType = jsonTypeSchema(types, type);
          if (schemaType) {
            acc.properties[name] = schemaType;
          }
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: []
        } as JsonSchemaObject & { required: string[] }
      );
    case "array":
      const itemsType = jsonTypeSchema(types, type.elements);
      if (!itemsType) {
        throw new Error(`Unsupported void array`);
      }
      return {
        type: "array",
        items: itemsType
      };
    case "optional":
      throw new Error(`Unsupported top-level optional type`);
    case "union":
      const unionTypes = type.types.map(t => jsonTypeSchema(types, t));
      const withoutNullTypes = compact(unionTypes);
      if (withoutNullTypes.length !== unionTypes.length) {
        throw new Error(`Unsupported void type in union`);
      }
      return {
        oneOf: withoutNullTypes
      };
    case "type-reference":
      return {
        $ref: `#/definitions/${type.typeName}`
      };
    default:
      throw assertNever(type);
  }
}

export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
