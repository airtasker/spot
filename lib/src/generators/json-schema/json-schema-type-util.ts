import assertNever from "assert-never";
import {
  areBooleanLiteralTypes,
  areFloatLiteralTypes,
  areIntLiteralTypes,
  areStringLiteralTypes,
  BooleanLiteralType,
  FloatLiteralType,
  IntLiteralType,
  isBooleanLiteralType,
  isFloatLiteralType,
  isIntLiteralType,
  isNotLiteralType,
  isStringLiteralType,
  StringLiteralType,
  Type,
  TypeKind
} from "../../types";
import { JsonSchemaObject, JsonSchemaType } from "./json-schema-specification";

/**
 * Generate a JSON Schema type definition. `objectAdditionalProperties` may
 * be used to configure whether additional properties should be allowed on
 * object types. This can be useful for data validation purposes where
 * property validation can be strict or lenient.
 *
 * @param type a contract type
 * @param objectAdditionalProperties whether to allow additional properties for objects
 */
export function typeToJsonSchemaType(
  type: Type,
  objectAdditionalProperties: boolean = true // TODO: expose proper configuration object
): JsonSchemaType {
  switch (type.kind) {
    case TypeKind.NULL:
      return {
        type: "null"
      };
    case TypeKind.BOOLEAN:
      return {
        type: "boolean"
      };
    case TypeKind.BOOLEAN_LITERAL:
      return {
        type: "boolean",
        const: type.value
      };
    case TypeKind.DATE:
      return { type: "string", format: "date" };
    case TypeKind.DATE_TIME:
      return { type: "string", format: "date-time" };
    case TypeKind.STRING:
      return {
        type: "string"
      };
    case TypeKind.STRING_LITERAL:
      return {
        type: "string",
        const: type.value
      };
    case TypeKind.FLOAT:
    case TypeKind.DOUBLE:
      return {
        type: "number"
      };
    case TypeKind.FLOAT_LITERAL:
      return {
        type: "number",
        const: type.value
      };
    case TypeKind.INT32:
    case TypeKind.INT64:
      return {
        type: "integer"
      };
    case TypeKind.INT_LITERAL:
      return {
        type: "integer",
        const: type.value
      };
    case TypeKind.OBJECT:
      return type.properties.reduce<JsonSchemaObject & { required: string[] }>(
        (acc, property) => {
          if (!property.optional) {
            acc.required.push(property.name);
          }
          acc.properties[property.name] = typeToJsonSchemaType(
            property.type,
            objectAdditionalProperties
          );
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: [],
          additionalProperties: objectAdditionalProperties
        }
      );
    case TypeKind.ARRAY:
      return {
        type: "array",
        items: typeToJsonSchemaType(
          type.elementType,
          objectAdditionalProperties
        )
      };
    case TypeKind.UNION:
      const elements = type.types;
      if (elements.length === 0) throw new Error("Union type has no elements");
      if (elements.length === 1) return typeToJsonSchemaType(elements[0]);

      if (
        areBooleanLiteralTypes(elements) ||
        areStringLiteralTypes(elements) ||
        areFloatLiteralTypes(elements) ||
        areIntLiteralTypes(elements)
      ) {
        return singleTypeLiteralsToSchema(elements);
      } else {
        // Guaranteed oneOf

        const oneOfElements = elements
          .filter(isNotLiteralType)
          .map(t => typeToJsonSchemaType(t, objectAdditionalProperties));

        const booleanLiterals = elements.filter(isBooleanLiteralType);
        if (booleanLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(booleanLiterals));
        }

        const stringLiterals = elements.filter(isStringLiteralType);
        if (stringLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(stringLiterals));
        }

        const floatLiterals = elements.filter(isFloatLiteralType);
        if (floatLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(floatLiterals));
        }

        const integerLiterals = elements.filter(isIntLiteralType);
        if (integerLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(integerLiterals));
        }

        return {
          oneOf: oneOfElements
        };
      }
    case TypeKind.REFERENCE:
      return {
        $ref: `#/definitions/${type.name}`
      };
    default:
      throw assertNever(type);
  }
}

function singleTypeLiteralsToSchema(
  literals:
    | BooleanLiteralType[]
    | StringLiteralType[]
    | FloatLiteralType[]
    | IntLiteralType[]
): JsonSchemaType {
  switch (literals.length) {
    case 0:
      throw new Error("no literals found");
    case 1:
      return typeToJsonSchemaType(literals[0]);
    default:
      if (areBooleanLiteralTypes(literals)) {
        literals.map(e => e.value);
        return {
          type: "boolean",
          enum: literals.map(e => e.value)
        };
      } else if (areStringLiteralTypes(literals)) {
        return {
          type: "string",
          enum: literals.map(e => e.value)
        };
      } else if (areFloatLiteralTypes(literals)) {
        return {
          type: "number",
          enum: literals.map(e => e.value)
        };
      } else if (areIntLiteralTypes(literals)) {
        return {
          type: "integer",
          enum: literals.map(e => e.value)
        };
      } else {
        throw new Error("Unknown literals");
      }
  }
}
