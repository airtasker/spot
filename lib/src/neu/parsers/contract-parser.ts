import { ClassDeclaration, SourceFile } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Config, Contract } from "../definitions";
import { ParserError } from "../errors";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { ok, Result } from "../util";
import { defaultConfig, parseConfig } from "./config-parser";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

/**
 * Parse a root source file to return a contract.
 */
export function parseContract(
  file: SourceFile
): Result<{ contract: Contract; lociTable: LociTable }, Error> {
  const typeTable = new TypeTable();
  const lociTable = new LociTable();

  const klass = getClassWithDecoratorOrThrow(file, "api"); // TODO: throw a custom error
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  // Handle config
  const configResult = resolveConfig(klass);
  if (configResult.isErr()) return configResult;

  // Handle security
  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");
  const security =
    securityHeaderProp &&
    parseSecurityHeader(securityHeaderProp, typeTable, lociTable);
  if (security && security.isErr()) return security;

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpointClasses = projectFiles.reduce<ClassDeclaration[]>(
    (acc, currentFile) =>
      acc.concat(
        currentFile
          .getClasses()
          .filter(k => k.getDecorator("endpoint") !== undefined)
      ),
    []
  );
  const endpoints = [];
  for (const k of endpointClasses) {
    const endpointResult = parseEndpoint(k, typeTable, lociTable);
    if (endpointResult.isErr()) return endpointResult;
    endpoints.push(endpointResult.unwrap());
  }

  const contract = {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    config: configResult.unwrap(),
    security: security && security.unwrap(),
    endpoints: endpoints.sort((a, b) => (b.name > a.name ? -1 : 1))
  };

  return ok({ contract, lociTable });
}

function resolveConfig(klass: ClassDeclaration): Result<Config, ParserError> {
  const hasConfigDecorator = klass.getDecorator("config") !== undefined;
  if (hasConfigDecorator) {
    return parseConfig(klass);
  } else {
    return ok(defaultConfig());
  }
}
