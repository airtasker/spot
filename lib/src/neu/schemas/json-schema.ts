export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
  additionalProperties: boolean;
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  format?: string;
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
