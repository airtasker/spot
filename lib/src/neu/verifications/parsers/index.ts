// @ts-ignore
import * as antlr4 from 'antlr4';
// @ts-ignore
import { QueryParamsLexer } from './QueryParamsLexer';
// @ts-ignore
import { QueryParamsParser } from './QueryParamsParser';

// let MyGrammarParser = require('./QueryParamsLexer').MyGrammarParser;
// let MyGrammarListener = require('./QueryParamsLexer').MyGrammarListener;

export function parseQueryParams(input:string) {
    let chars = new antlr4.InputStream(input);
    let lexer = new QueryParamsLexer(chars);
    let tokens  = new antlr4.CommonTokenStream(lexer);
    // console.log(tokens);
    let parser = new QueryParamsParser(tokens);
    parser.buildParseTrees = true;
    let tree = parser.querystring();
    console.log(tree.toStringTree(parser.ruleNames)); 
}
