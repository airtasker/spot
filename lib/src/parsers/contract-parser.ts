import { ClassDeclaration, SourceFile } from "ts-morph";
import { Config, Contract, Endpoint } from "../definitions";
import { ParserError } from "../errors";
import { LociTable } from "../locations";
import { ApiConfig } from "../syntax/api";
import { TypeTable } from "../types";
import { err, ok, Result } from "../util";
import { defaultConfig, parseConfig } from "./config-parser";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralProp,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

import { parseOa3Servers } from "./oa3server-parser";

/**
 * Parse a root source file to return a contract.
 */
export function parseContract(
  file: SourceFile
): Result<{ contract: Contract; lociTable: LociTable }, ParserError> {
  const typeTable = new TypeTable();
  const lociTable = new LociTable();

  const klass = getClassWithDecoratorOrThrow(file, "api");
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);

  // Handle name
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const name = nameLiteral.getLiteralText().trim();
  if (name.length === 0) {
    return err(
      new ParserError("api name cannot be empty", {
        file: nameLiteral.getSourceFile().getFilePath(),
        position: nameLiteral.getPos()
      })
    );
  }
  if (!/^[\w\s-]*$/.test(name)) {
    return err(
      new ParserError(
        "api name may only contain alphanumeric, space, underscore and hyphen characters",
        {
          file: nameLiteral.getSourceFile().getFilePath(),
          position: nameLiteral.getPos()
        }
      )
    );
  }

  // Handle description
  const descriptionDoc = getJsDoc(klass);
  const description = descriptionDoc?.getDescription().trim();

  // Handle Version
  const versionProp = getObjLiteralProp<ApiConfig>(decoratorConfig, "version");
  const version = versionProp
    ? getPropValueAsStringOrThrow(versionProp).getLiteralText().trim()
    : undefined;

  // Handle config
  const configResult = resolveConfig(klass);
  if (configResult.isErr()) return configResult;
  const config = configResult.unwrap();

  // Handle security
  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");
  const securityResult =
    securityHeaderProp &&
    parseSecurityHeader(securityHeaderProp, typeTable, lociTable);
  if (securityResult && securityResult.isErr()) return securityResult;
  const security = securityResult?.unwrap();

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpointClasses = projectFiles.reduce<ClassDeclaration[]>(
    (acc, currentFile) =>
      acc.concat(
        currentFile
          .getClasses()
          .filter(k => k.getDecorator("endpoint") !== undefined)
      ),
    []
  );
  const endpointsResult = extractEndpoints(
    endpointClasses,
    typeTable,
    lociTable
  );
  if (endpointsResult.isErr()) return endpointsResult;
  const endpoints = endpointsResult.unwrap();

  // Handle Types
  const types = typeTable.toArray();

  // Handle Servers
  const serversResult = parseOa3Servers(klass, typeTable, lociTable);
  if (serversResult.isErr()) return serversResult;
  const oa3servers = serversResult.unwrap();

  const contract = {
    name,
    description,
    types,
    config,
    security,
    endpoints,
    version,
    oa3servers
  };
  return ok({ contract, lociTable });
}

function resolveConfig(klass: ClassDeclaration): Result<Config, ParserError> {
  const hasConfigDecorator = klass.getDecorator("config") !== undefined;
  if (hasConfigDecorator) {
    return parseConfig(klass);
  } else {
    return ok(defaultConfig());
  }
}

function extractEndpoints(
  endpointClasses: ClassDeclaration[],
  typeTable: TypeTable,
  lociTable: LociTable
): Result<Endpoint[], ParserError> {
  const endpointNames = endpointClasses.map(k => k.getNameOrThrow());
  const duplicateEndpointNames = [
    ...new Set(
      endpointNames.filter(
        (name, index) => endpointNames.indexOf(name) !== index
      )
    )
  ];
  if (duplicateEndpointNames.length !== 0) {
    const locations = duplicateEndpointNames.reduce<
      { file: string; position: number }[]
    >((acc, name) => {
      const nameLocations = endpointClasses
        .filter(k => k.getNameOrThrow() === name)
        .map(k => {
          return {
            file: k.getSourceFile().getFilePath(),
            position: k.getPos()
          };
        });
      return acc.concat(nameLocations);
    }, []);

    return err(
      new ParserError("endpoints must have unique names", ...locations)
    );
  }

  const endpoints = [];
  for (const k of endpointClasses) {
    const endpointResult = parseEndpoint(k, typeTable, lociTable);
    if (endpointResult.isErr()) return endpointResult;
    endpoints.push(endpointResult.unwrap());
  }
  return ok(endpoints);
}
