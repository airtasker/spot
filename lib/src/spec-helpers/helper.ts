import * as path from "path";
import { Project, SourceFile, ts } from "ts-morph";

/**
 * Create an AST source file. Any files imported from the main file must also be provided.
 * All files will be loaded into a virtual filesystem under a `test/` directory.
 *
 * @param mainFile details for main file
 * @param referencedContent details for referenced files
 * @returns the main source file
 */
export function createSourceFile(
  mainFile: FileDetail,
  ...referencedFiles: FileDetail[]
): SourceFile {
  const project = createProject();
  referencedFiles.forEach(fileDetail => {
    project.createSourceFile(`test/${fileDetail.path}.ts`, fileDetail.content);
  });
  const mainSource = project.createSourceFile(
    `test/${mainFile.path}.ts`,
    mainFile.content
  );

  validateProject(project);

  return mainSource;
}

interface FileDetail {
  /** File path */
  path: string;
  /** File content */
  content: string;
}

/**
 * Create an AST project with the `@airtasker/spot` dependency loaded.
 */
export function createProject(): Project {
  return new Project({
    compilerOptions: {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.CommonJS,
      strict: true,
      noImplicitAny: true,
      strictNullChecks: true,
      strictFunctionTypes: true,
      strictPropertyInitialization: true,
      noImplicitThis: true,
      alwaysStrict: true,
      noImplicitReturns: true,
      noFallthroughCasesInSwitch: true,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
      experimentalDecorators: true,
      baseUrl: "./",
      paths: {
        "@airtasker/spot": [path.join(__dirname, "../lib")]
      }
    }
  });
}

export function createProjectFromExistingSourceFile(
  filePath: string
): { project: Project; file: SourceFile } {
  const project = createProject();
  const file = project.addSourceFileAtPath(filePath);
  project.resolveSourceFileDependencies();
  validateProject(project);
  return { project, file };
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
export function validateProject(project: Project): void {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
