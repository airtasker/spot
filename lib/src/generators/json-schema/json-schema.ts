import { Contract } from "../../definitions";
import { JsonSchema, JsonSchemaType } from "./json-schema-specification";
import { typeToJsonSchemaType } from "./json-schema-type-util";

export function generateJsonSchema(contract: Contract): JsonSchema {
  return {
    $schema: "http://json-schema.org/draft-07/schema#",
    definitions: contract.types.reduce<{
      [typeName: string]: JsonSchemaType;
    }>((acc, typeNode) => {
      acc[typeNode.name] = typeToJsonSchemaType(typeNode.typeDef.type);
      return acc;
    }, {})
  };
}
