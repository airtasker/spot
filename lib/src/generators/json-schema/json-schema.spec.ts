import { isJSONSchemaDraft7, Spectral } from "@stoplight/spectral";
import { parseContract } from "../../parsers/contract-parser";
import { createProjectFromExistingSourceFile } from "../../spec-helpers/helper";
import { generateJsonSchema } from "./json-schema";

describe("JSON Schema generator", () => {
  const spectral = new Spectral();

  beforeAll(async () => {
    spectral.registerFormat("json-schema-draft7", isJSONSchemaDraft7);
  });

  test("produces minimal json schema", async () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/minimal-contract.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateJsonSchema(contract);

    expect(result.$schema).toEqual("http://json-schema.org/draft-07/schema#");
    expect(result.definitions).toEqual({});
    expect(result).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });

  test("produces definitions", async () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/contract-with-reference-types.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateJsonSchema(contract);

    expect(result.definitions).toHaveProperty("User");
    expect(result.definitions).toHaveProperty("Users");
    expect(result).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });

  test("evaluates intersection type", async () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/contract-with-intersection-types.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateJsonSchema(contract);

    expect(result).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });
});
