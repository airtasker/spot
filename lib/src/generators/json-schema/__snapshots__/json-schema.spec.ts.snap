// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator evaluates intersection type 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "IntersectionResponse": Object {
      "allOf": Array [
        Object {
          "additionalProperties": true,
          "properties": Object {
            "id": Object {
              "type": "string",
            },
          },
          "required": Array [
            "id",
          ],
          "type": "object",
        },
        Object {
          "additionalProperties": true,
          "properties": Object {
            "name": Object {
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        },
      ],
    },
  },
}
`;

exports[`JSON Schema generator produces definitions 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "User": Object {
      "additionalProperties": true,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "name",
      ],
      "type": "object",
    },
    "Users": Object {
      "items": Object {
        "$ref": "#/definitions/User",
      },
      "type": "array",
    },
  },
}
`;

exports[`JSON Schema generator produces minimal json schema 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {},
}
`;
